<?xml version="1.0"?>
<launch>

    <!--initialize the arguments-->

    <arg name="world_number" default="0"/>
    
    <arg name="gui" default="true" />
    <arg name="mpc_model" default="0"/>
    <arg name="enable_dynamic_obstacles" default="false"/>
    <arg name="enable_static_obstacles" default="false"/>
    <arg name="output" default="screen"/>

    <arg name="world_name" if="$(eval arg('world_number')==0)" value="track_porto"/>
    <arg name="world_name" if="$(eval arg('world_number')==1)" value="racecar_walker"/>
    <arg name="world_name" if="$(eval arg('world_number')==2)" value="track_barca"/>

    <!--Initial States of Vehicles for this scenario, ego car is racecar2
        The intitial positions also depend on the racetrack-->
    <arg name="racecar_pose" default="-x -3.5 -y -0.2 -z 0.05"  if="$(eval arg('world_name')=='track_porto')"/>
    <arg name="racecar2_pose" default="-x 1.5 -y 1.0 -z 0.05"   if="$(eval arg('world_name')=='track_porto')"/>
    <arg name="racecar3_pose" default="-x -1.5 -y 0.05 -z 0.05" if="$(eval arg('world_name')=='track_porto')"/>

    <arg name="racecar_pose" default="-x 0.0898 -y -0.174 -z 0.05"  if="$(eval arg('world_name')=='track_barca')"/>
    <arg name="racecar2_pose" default="-x -2.68 -y 1.38 -z 0.05"   if="$(eval arg('world_name')=='track_barca')"/>
    <arg name="racecar3_pose" default="-x -7.63 -y 3.61 -z 0.05" if="$(eval arg('world_name')=='track_barca')"/>

    <arg name="racecar_pose" default="-x 4.03 -y 0.324 -z 0.05"  if="$(eval arg('world_name')=='racecar_walker')"/>
    <arg name="racecar2_pose" default="-x -0.219 -y 0.376 -z 0.05"   if="$(eval arg('world_name')=='racecar_walker')"/>
    <arg name="racecar3_pose" default="-x -3.5 -y -0.711 -z 0.05" if="$(eval arg('world_name')=='racecar_walker')"/>

    <!--waypoint file used by pure pursuit algorithm-->
    <arg name="csv_filename" default="track_porto_26780.csv"    if="$(eval arg('world_name')=='track_porto')"/>
    <arg name="csv_filename" default="racecar_walker_28998.csv" if="$(eval arg('world_name')=='racecar_walker')"/>
    <arg name="csv_filename" default="track_barca_47396.csv"    if="$(eval arg('world_name')=='track_barca')"/>

    <arg name="map_file" default="$(find race)/maps/porto_track.yaml"  if="$(eval arg('world_name')=='track_porto')"/>
    <arg name="map_file" value="$(find race)/maps/racecar_walker.yaml" if="$(eval arg('world_name')=='racecar_walker')"/>
    <arg name="map_file"  value="$(find race)/maps/track_barca.yaml"   if="$(eval arg('world_name')=='track_barca')"/> 

    <!--Handles finding the wall points-->
    <arg name="wall_points" if="$(eval arg('world_name')=='track_porto')" value="porto_obstacles.txt"/>
    <arg name="wall_points" if="$(eval arg('world_name')=='racecar_walker')" value="racecar_walker_obstacles.txt"/>
    <arg name="wall_points" if="$(eval arg('world_name')=='track_barca')" value="track_barca_obstacles.txt"/>

    <!--random seed used to allocte obstacles within vehicle environment-->
    <arg name="random_seed" default = "1516"/>

    <!--docker requirements-->
    <arg name="use_sim_time" default = "true"/>
    <arg name="verbose" default ="false"/>

    <!--multi-agent parameters-->
    <!--How many boxes to let rviz visualize-->
    <arg name="box_display_limit" default="100"/>

    <!-- reach-time is the time we are computing reachability for-->
    <arg name="reach_time" default="0.5"/>

    <!--wall-time is the time we allow the reachability algorithm to run for-->
    <arg name="wall_time" default="1.0"/>

    <!--number of vehicles to launch in the simulation environment-->
    <arg name='number_of_cars' default='2'/>

    <!--Select the car we are computing reachability for-->
    <arg name='car_name' default='racecar2'/>

    <arg name="rviz_disp" default="multiagent_reach.rviz"/>


    <!--  This is the main launch file that launches the world, plugins, controllers, vehicles -->
    <include file="$(find race)/launch/multi_agent_gen.launch">
      <arg name="racecar_pose" value="$(arg racecar_pose)"/>
      <arg name="racecar2_pose" value="$(arg racecar2_pose)"/>
      <arg name="racecar3_pose" value="$(arg racecar3_pose)"/>
      <arg name="number_of_cars" value="$(arg number_of_cars)"/>
      <arg name="gui" value="$(arg gui)"/>
      <arg name="rviz_disp" value="$(arg rviz_disp)"/>
      <arg name="world_name" value="$(arg world_name)"/>
    </include>

    <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_file)" />

    <!--
      GPU based particle filtering from MIT
    
     <include file="$(find particle_filter)/launch/localize.launch">
      <arg name="scan_topic" default="racecar/scan"/>
	    <arg name="odometry_topic" default="racecar/odom"/>
      <arg name="laser_frame" default="racecar/laser"/>
    </include>
    -->

    <node pkg="mpc" type="mpc_drive_hyperplanes.py" name="mpc_node" output="$(arg output)" if="$(eval arg('mpc_model')==0)" 
          args="racecar2 $(arg csv_filename) $(arg wall_points)" required="true"/> 
          
    <node pkg="mpc" type="mpcc_hyperplanes.py" name="mpc_node" output="$(arg output)" 
                    if="$(eval arg('mpc_model')==1)" args="racecar2 $(arg csv_filename) $(arg wall_points)" 
                    required="true"/>

    <!--Uncomment this if you want to run another vehicle with the MPCC formulation-->
    <!-- <node pkg="mpc" type="mpcc_hyperplanes.py" name="mpc_node2" output="$(arg output)" if="$(eval arg('mpc_model')==0)" args="racecar" required="true"/> -->


    <!--Launch Pure Pursuit Node-->
    <!-- <node pkg="pure_pursuit" type="pure_pursuit_angle.py" name="pure_pursuit" args="$(arg car_name) $(arg csv_filename) none" output="$(arg output)" required="true"/> -->

    <!--Launch Disparity Extender Waypoint Following-->
    <node pkg="race" name="disparity_extender_target_point" type="disparity_extender_ftg.py" args="racecar2" required="true"/>
    <!-- <node pkg="race" name="disparity_extender_target_point1" type="disparity_extender_ftg.py" args="racecar" required="true"/> -->

    <node name="vehicle_1" pkg="rtreach" type="visualize_obs" args="racecar   $(arg wall_time) $(arg reach_time) $(arg box_display_limit)" output="log"/>
    <node name="vehicle_2" pkg="rtreach" type="visualize_obs" args="racecar3   $(arg wall_time) $(arg reach_time) $(arg box_display_limit)" output="log"/>

    <!-- controller for opponent vehicle  -->
    <node pkg="race" name="disparity_extender_opponent_vehicle" type="disparity_extender_vanderbilt_gen.py" args="racecar 30"/>
    <!-- <node pkg="race" name="disparity_extender_opponent_vehicle1" type="disparity_extender_vanderbilt_gen.py" args="racecar2 10"/> -->
    

    <!-- controller for opponent vehicle  -->
    <!-- <node pkg="race" name="disparity_extender_opponent_vehicle2" type="disparity_extender_vanderbilt_gen.py" args="racecar3 15"/> -->

    <!-- <node pkg="pure_pursuit" type="pure_pursuit.py" name="pure_pursuit_opp_car2" args="racecar $(arg csv_filename)" output="$(arg output)" required="true"/> 
    <node pkg="pure_pursuit" type="pure_pursuit.py" name="pure_pursuit_opp_car" args="racecar3 $(arg csv_filename)" output="$(arg output)" required="true"/>  -->

    <!--Spawn Cones-->
    <node pkg="racecar_gazebo" name="spawn_cones" type="spawn_zigzag.py" if="$(arg enable_static_obstacles)" output= "screen" required="true"/>

    <include file="$(find mpc)/launch/dynamic_boxes_launch.launch" if="$(arg enable_dynamic_obstacles)"/>
</launch>
